### PR TITLE
[5.8] Fix for container app()->call() method

### PR DESCRIPTION
**Pull request by imanghafoori1**
Addresses these issues :
https://github.com/laravel/framework/issues/26851
https://github.com/laravel/framework/issues/22687
https://github.com/laravel/framework/issues/26837

Following my previous bug fix https://github.com/laravel/framework/pull/26852 which caused a new nasty bug and eventually got reverted...
I created an other PR to add the missing tests to cover more cases https://github.com/laravel/framework/pull/26897

Now I am coming up with an other solution for making it really fixed.

### CODE REVIEW

**Comment by imanghafoori1** [comment]
Some changes are accidentally made by IDE.
But this fix required a lot of change in the logic.
I try to provide enough comments, meaningful variable names and unit test to be clear.

Anyway this is a work in progress and still requires a lot more consideration.

File: src/Illuminate/Container/BoundMethod.php [hunk 1]
```diff
@@ -105,18 +105,25 @@ protected static function normalizeMethod($callback)
...
-        foreach (static::getCallReflector($callback)->getParameters() as $parameter) {
-            static::addDependencyForCallParameter($container, $parameter, $parameters, $dependencies);
+        // In case the method has no explicit input parameters defined
+        // we will call it, with whatever input data available to us.
+        if (count($signature) === 0) {
+            return $inputData;
         }
 
-        return array_merge($dependencies, $parameters);
+        if (\Illuminate\Support\Arr::isAssoc($inputData)) {
```
Comment (laurencei) [inline src/Illuminate/Container/BoundMethod.php]: You should import this at the top, and just use `Arr` here.

File: src/Illuminate/Container/BoundMethod.php [hunk 2]
```diff
@@ -127,42 +134,47 @@ protected static function getMethodDependencies($container, $callback, array $pa
...
+            if (array_key_exists($parameter->name, $inputData)) {
+                $resolvedInputData[] = $inputData[$parameter->name];
+            } elseif ($parameter->getClass() && array_key_exists($parameter->getClass()->name, $inputData)) {
+                $resolvedInputData[] = $inputData[$parameter->getClass()->name];
+            } elseif ($parameter->getClass()) {
+                $resolvedInputData[] = $container->make($parameter->getClass()->name);
+            } elseif (isset($inputData[$i])) {
+                $data = $inputData[$i];
+                $i++;
+                $resolvedInputData[] = $data;
```
Comment (laurencei) [inline src/Illuminate/Container/BoundMethod.php]: I think you can just do `$resolvedInputData[] = $inputData[$i]` here and not have `$data`?

Comment (imanghafoori1) [reply src/Illuminate/Container/BoundMethod.php]: That part will change, although all tests are passing by now, I am pretty well sure that there is a bug sneaking us.

File: src/Illuminate/Container/BoundMethod.php (hunk 1 above)
Comment (imanghafoori1) [reply src/Illuminate/Container/BoundMethod.php]: I have no idea about the standard conventions of laravel for it.
It can also be imported.

**Review by jmarcher — APPROVED** [review]

File: src/Illuminate/Container/BoundMethod.php [hunk 3]
```diff
@@ -105,18 +106,25 @@ protected static function normalizeMethod($callback)
...
      */
-    protected static function getMethodDependencies($container, $callback, array $parameters = [])
+    protected static function getMethodDependencies($container, $callback, array $inputData = []) : array
     {
-        $dependencies = [];
+        $signature = static::getCallReflector($callback)->getParameters();
+
+        // In case the method has no explicit input parameters defined
+        // we will call it, with whatever input data available to us.
+        if (count($signature) === 0) {
```
Comment (mfn) [inline src/Illuminate/Container/BoundMethod.php]: Is this to support `func_get_args()` in the called method? If so, I would like the comments to reflect this. Thanks

**Comment by mfn** [comment]
Thanks for the PR, I like it 👍

File: src/Illuminate/Container/BoundMethod.php (hunk 3 above)
Comment (imanghafoori1) [reply src/Illuminate/Container/BoundMethod.php]: sure

File: src/Illuminate/Container/BoundMethod.php [hunk 4]
```diff
@@ -105,18 +106,25 @@ protected static function normalizeMethod($callback)
      *
      * @param  \Illuminate\Container\Container  $container
      * @param  callable|string  $callback
-     * @param  array  $parameters
+     * @param  array $inputData
      * @return array
+     * @throws \ReflectionException
      */
-    protected static function getMethodDependencies($container, $callback, array $parameters = [])
+    protected static function getMethodDependencies($container, $callback, array $inputData = []) : array
```
Comment (laurencei) [inline src/Illuminate/Container/BoundMethod.php]: why did you change `$parameters` to `$inputData`? Cant the name stay as it was to reduce unnecessary changes?

Comment (imanghafoori1) [reply src/Illuminate/Container/BoundMethod.php]: of course why not ? the term 'parameter' was overloaded here and it had 2 meanings.
You know it confused me at first, which $parameter refers to the actual user input and which one is the method signature declaration. So I wanted to make it crystal clear

File: src/Illuminate/Container/BoundMethod.php [hunk 5]
```diff
@@ -106,11 +105,12 @@ protected static function normalizeMethod($callback)
      *
      * @param  \Illuminate\Container\Container  $container
      * @param  callable|string  $callback
-     * @param  array $inputData
+     * @param  array  $inputData
      * @return array
+     *
```
Comment (driesvints) [inline src/Illuminate/Container/BoundMethod.php]: @GrahamCampbell we don't add these white lines above `@throws` anymore: https://laravel.com/docs/5.7/contributions#coding-style

Comment (imanghafoori1) [reply src/Illuminate/Container/BoundMethod.php]: @driesvints Is there any convention for avoiding php 7 return types?

Comment (X-Coder264) [reply src/Illuminate/Container/BoundMethod.php]: @imanghafoori1 The PHP 7 type-hinting is being discussed here: https://github.com/laravel/ideas/issues/1409

**Comment by driesvints** [comment]
@imanghafoori1 your branch conflicts with the master branch. You'll need to fix the conflicts.

**Comment by imanghafoori1** [comment]
It seems to be done from my point of view, but I need other pair of careful eyes, to review and comment it.

**Comment by driesvints** [comment]
@imanghafoori1 sorry, the Github GUI showed me some rebase notices but those turned out to be unrelated of conflicts.

**Comment by imanghafoori1** [comment]
Ok deleting extra comments related to conflict.

**Comment by imanghafoori1** [comment]
It seems the ContainerTest.php file needs to be split apart into more files, each covering one specific feature. I do not know whether it is against any convention or not.
(For Example all the tests covering the app()->call() should reside in one file.)

File: src/Illuminate/Container/BoundMethod.php (hunk 5 above)
Comment (GrahamCampbell) [reply src/Illuminate/Container/BoundMethod.php]: @driesvints Oh, that's odd, because every throws annotation in the entire codebase has a line above it (bar 1 exception that I noticed, looking through the search results). Are the docs wrong?

image

File: src/Illuminate/Container/BoundMethod.php [hunk 6]
```diff
@@ -5,71 +5,78 @@
 use Closure;
 use ReflectionMethod;
 use ReflectionFunction;
+use Illuminate\Support\Arr;
 use InvalidArgumentException;
 
 class BoundMethod
 {
+    /**
+     * @var \Illuminate\Foundation\Application
```
Comment (X-Coder264) [inline src/Illuminate/Container/BoundMethod.php]: This should be `\Illuminate\Contracts\Container\Container`

The `illuminate/container` package can't reference something from the `Foundation` namespace.

File: src/Illuminate/Container/BoundMethod.php [hunk 7]
```diff
@@ -139,40 +157,71 @@ protected static function getCallReflector($callback)
...
+            return $inputData[$i++];
+        } else {
+            return self::$container->make($className);
+        }
+    }
+
+    /**
+     * @param $container
+     */
+    public static function setContainer($container)
```
Comment (X-Coder264) [inline src/Illuminate/Container/BoundMethod.php]: Add the `Illuminate\Contracts\Container\Container` typehint.

Comment (X-Coder264) [reply src/Illuminate/Container/BoundMethod.php]: I meant adding the actual typehint, not just to the docblock :)

File: src/Illuminate/Container/BoundMethod.php [hunk 8]
```diff
@@ -5,71 +5,78 @@
...
 use ReflectionFunction;
+use Illuminate\Support\Arr;
 use InvalidArgumentException;
 
 class BoundMethod
 {
+    /**
+     * @var \Illuminate\Contracts\Container\Container
+     */
+    private static $container;
```
Comment (X-Coder264) [inline src/Illuminate/Container/BoundMethod.php]: What is the need here for memoization? This PR should be focused on fixing the issue at hand, memoization is not needed for that. You had to change some additional method signatures to accomplish that which means more breaking changes than what is needed to fix the problem that this PR is trying to solve. The memoization should be a separate PR. And what are the performance gains here (Taylor usually doesn't merge PRs which supposedly improve the performance of some part of the framework without some benchmarks)?

File: src/Illuminate/Container/BoundMethod.php [hunk 9]
```diff
@@ -139,40 +158,73 @@ protected static function getCallReflector($callback)
...
+        } else {
+            // resolves a new instance
+            return self::$container->make($className);
+        }
+    }
+
+    /**
+     * @param ContainerContract $container
+     */
+    public static function setContainer(ContainerContract $container)
```
Comment (X-Coder264) [inline src/Illuminate/Container/BoundMethod.php]: This should be nullable as well as you are passing null here in some cases (the same applies to the `call` method).

**Comment by imanghafoori1** [comment]
Ok, hope I have manged to provide something that can be inspired from in future.

**Comment by imanghafoori1** [comment]
@taylorotwell 
I see that if this issue gets fixed, then laravel can use $app->call(... in the base facade class to call the underlying implementation of facades.

This means that the end user can enjoy the automatic method injection when using real-time facades.
